src/OpenTelemetry.Exporter.Prometheus/PrometheusExporterMetricsHttpServer.cs

```diff
@@ -0,0 +1,145 @@
+        {
+            lock (this.syncObject)
+            {
+                if (this.tokenSource != null)
+                {
+                    return;
+                }
+
+                // link the passed in token if not null
+                this.tokenSource = token == default ?
+                    new CancellationTokenSource() :
+                    CancellationTokenSource.CreateLinkedTokenSource(token);
+
+                this.workerThread = Task.Factory.StartNew(this.WorkerThread, default, TaskCreationOptions.LongRunning, TaskScheduler.Default);
+            }
+        }
+
+        /// <summary>
+        /// Stop exporter.
+        /// </summary>
+        public void Stop()
+        {
+            lock (this.syncObject)
+            {
+                if (this.tokenSource == null)
+                {
+                    return;
+                }
+
+                this.tokenSource.Cancel();
+                this.workerThread.Wait();
+                this.tokenSource = null;
+            }
+        }
+
+        /// <inheritdoc/>
+        public void Dispose()
+        {
+            this.Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        /// <summary>
+        /// Releases the unmanaged resources used by this class and optionally releases the managed resources.
+        /// </summary>
+        /// <param name="disposing"><see langword="true"/> to release both managed and unmanaged resources; <see langword="false"/> to release only unmanaged resources.</param>
+        protected virtual void Dispose(bool disposing)
+        {
+            if (this.httpListener != null && this.httpListener.IsListening)
+            {
+                this.Stop();
+                this.httpListener.Close();
+            }
+        }
+
+        private void WorkerThread()
+        {
+            this.httpListener.Start();
+
+            try
+            {
+                while (!this.tokenSource.IsCancellationRequested)
+                {
+                    var ctxTask = this.httpListener.GetContextAsync();
+                    ctxTask.Wait(this.tokenSource.Token);
+
+                    var ctx = ctxTask.Result;
+
+                    ctx.Response.StatusCode = 200;
+                    ctx.Response.ContentType = PrometheusMetricBuilder.ContentType;
+
+                    using var output = ctx.Response.OutputStream;
+                    using var writer = new StreamWriter(output);
+                    this.exporter.MakePullRequest();
+                    this.exporter.WriteMetricsCollection(writer);
+                }
+            }
+            catch (OperationCanceledException ex)
+            {
+                PrometheusExporterEventSource.Log.CanceledExport(ex);
+            }
+            catch (Exception ex)
+            {
+                PrometheusExporterEventSource.Log.FailedExport(ex);
+            }
+            finally
+            {
+                this.httpListener.Stop();
```

Comment:
Is it possible for these two lines to throw, and if yes what would happen to the user app?